File: Python/Array/The Number of Weak Characters in the Game.py
#Problem link:https://leetcode.com/problems/the-number-of-weak-characters-in-the-game/

#Time Complexity: O(N logN)
#Space Complexity: O(1)

class Solution:
    def numberOfWeakCharacters(self, properties: List[List[int]]) -> int:
        
        #number to be returned
        numberOfWeakCharacters = 0
        
        #record maximum defense
        maxDefenseScore = 0
        
        #sort the properties
        #ascending order for attacking scores
        #if attack is the same
        #descending order for defending scores
        properties.sort(key = lambda x: (x[0], -x[1]), reverse = True)
        
        #we only care about the defense score
        for _, defense in properties:
            
            #if less than maxDefenseScore then is a weak character
            if defense < maxDefenseScore:
                numberOfWeakCharacters +=1
            else:#update max defense
                maxDefenseScore = defense
        
        return numberOfWeakCharacters

File: Java/Array/The Number of Weak Characters in the Game.java
// problem Link: https://leetcode.com/problems/the-number-of-weak-characters-in-the-game/

/****

Approach: sort the array in decreasing order of attacks and 
in increasing order of defence.

Finally, check if value of defence is less than or not of current_max_defense value

 */


/***

Time Complexity: O(N logN)
Space Complexity: O(1)

 */


class Solution {
    public int numberOfWeakCharacters(int[][] arr) {
        int max=0,count=0;
        Arrays.sort(arr, (a,b) -> (b[0]==a[0])?a[1]-b[1]:b[0]-a[0]);
        for (int i=0;i<arr.length;i++){
            if(arr[i][1]<max){
                count++;
            }
            max=(int)Math.max(max,arr[i][1]);
        }
        return count;
    }
}